x-pack/plugins/security/common/licensing/license_service.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Observable, Subscription, of } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { ILicense } from '../../../licensing/common/types';
+import { SecurityLicenseFeatures } from './license_features';
+
+export interface SecurityLicense {
+  isEnabled(): boolean;
+  getFeatures(): SecurityLicenseFeatures;
+  features$: Observable<SecurityLicenseFeatures>;
+}
+
+interface SetupDeps {
+  license$: Observable<ILicense>;
+}
+
+export class SecurityLicenseService {
+  private licenseSubscription?: Subscription;
+
+  public setup({ license$ }: SetupDeps) {
+    let rawLicense: Readonly<ILicense> | undefined;
+
+    this.licenseSubscription = license$.subscribe(nextRawLicense => {
+      rawLicense = nextRawLicense;
+    });
+
+    return {
+      license: Object.freeze({
+        isEnabled: () => this.isSecurityEnabledFromRawLicense(rawLicense),
+
+        getFeatures: () => this.calculateFeaturesFromRawLicense(rawLicense),
+
+        features$: license$.pipe(
+          map(nextRawLicense => this.calculateFeaturesFromRawLicense(nextRawLicense))
+        ),
+      }),
+    };
+  }
+
+  public stop() {
+    if (this.licenseSubscription) {
+      this.licenseSubscription.unsubscribe();
+      this.licenseSubscription = undefined;
+    }
+  }
+
+  private isSecurityEnabledFromRawLicense(rawLicense: Readonly<ILicense> | undefined) {
+    if (!rawLicense) {
+      return false;
+    }
+
+    const securityFeature = rawLicense.getFeature('security');
+    return (
+      securityFeature !== undefined && securityFeature.isAvailable && securityFeature.isEnabled
+    );
+  }
+
+  private calculateFeaturesFromRawLicense(
+    rawLicense: Readonly<ILicense> | undefined
+  ): SecurityLicenseFeatures {
+    // If, for some reason, we cannot get license information from Elasticsearch,
+    // assume worst-case and lock user at login screen.
+    if (rawLicense === undefined || !rawLicense.isAvailable) {
+      return {
+        showLogin: true,
+        allowLogin: false,
+        showLinks: false,
+        allowRoleDocumentLevelSecurity: false,
+        allowRoleFieldLevelSecurity: false,
+        allowRbac: false,
+        layout:
```

Comment:
__nit:__ (I admit we need time to get used to these `?` everywhere)
```suggestion
        layout: !rawLicense?.isAvailable ? 'error-xpack-unavailable' : 'error-es-unavailable',
```

Reply:
I had to tweak this in https://github.com/elastic/kibana/pull/52386/commits/03a1409d699b48520b4568fe28a11d6c711cd5ff due to a test failure. We were relying on the distinction between `rawLicense` being undefined and `rawLicense.isAvailable` being falsy